### ISSUE TITLE
digicam scripts - command line args

### ISSUE THREAD

**Issue by dneise** [calispac/digicampipe#74]
There are some scripts in `digicampipe/` which cannot be used without modification, as for example they contain fixed paths into somebodies home folder. This should be avoided.

I think it can only be avoided using command line arguments. 
I know about 3 packages which are commonly used for this:
 - [argparse in the stdlib](https://docs.python.org/3/library/argparse.html)
 - [docopt](http://docopt.org/) (also available for C++ as a header-only library, really nice)
 - [click](http://click.pocoo.org/5/)
 
But there is many more:
https://awesome-python.com/#command-line-tools

We should choose one, and use that inside digicampipe.

**Comment by dneise** [comment]
### Input paths

When providing input paths for a program, many ways are possible. In some of the scripts I have seen this option:
```python
    directory = '/home/alispach/data/CRAB_01/'
    filename = directory + 'CRAB_01_0_000.%03d.fits.fz'
    file_list = [filename % number for number in range(3, 23)]
```

One could now think about command line arguments like these:
(using docopt style)
```
Usage:
    scriptname [options]

Options:
-d <dir>, --directory=<dir>  input directory [default:  '/home/alispach/data/CRAB_01/']
--filename=<filename>        template of the filename to be looked for [default:  'CRAB_01_0_000.%03d.fits.fz']
--start=<integer>            first run to be analyzed [default: 3]
--end=<integer>              last run to be analyzed [default: 23]
```

**But** ... this does not follow the linux philosophy. We should not re-implement functionality other people have already implemented for us. In this case it is the bash (or any other shell).

The linux bash (and also the other shells like zsh, tcsh)
is very powerful when it comes to files paths. In fact, this is what it was written for.

So our programs should not reproduce functionality, already implemented very well 
into the shell.

Consider this program call:

    echo foo*.txt

In a folder containing these files:

    foo_1.txt
    foo_2.txt
    foo_3.txt

What does the program `echo` see as its cli arguements? Well, the bash interprets 
the globbing character `*` and invokes `echo` as if the user had typed this:

    echo foo_1.txt foo_2.txt foo_3.txt 

So the user does not have to type in a lot of paths, but the program 
sees a lot of paths. 

Assuming a folder with 100 files with names `foo_000.txt` until `foo_099.txt`
and a user who only wants to "process" 20 files with names from `030` until `049`
might type this:

    echo foo_0[34]?.txt

Which expands to:
    
    echo foo_030.txt foo_031.txt foo_032.txt foo_033.txt foo_034.txt \
        foo_035.txt foo_036.txt foo_037.txt foo_038.txt foo_039.txt \
        foo_040.txt foo_041.txt foo_042.txt foo_043.txt foo_044.txt \
        foo_045.txt foo_046.txt foo_047.txt foo_048.txt foo_049.txt


-----

So coming back to our original question what command line interface should our script have?
I think this is a nice one:
```
Usage:
    scriptname <outputfile> <file>...
```

No options ... just a clean interface. Users *have to* specify where the output should go. 
And they **have to** clearly give a list of input files. 

So there is no awkward situation like: 

- "Okay, I executed that script, but I have no idea if it worked. I don't see any output." 
- "Ah yes, you have to look in this-and-that folder of course."
- "Yes of course :-P".

This interface I propose up there is very similar to the interface of programs people know already, that do similar things. An example would be [`tar`](https://linux.die.net/man/1/tar). `tar` also can get a list of files, to compress and pack into one output file. This is in a sense similar to an analysis of a bunch of runs. It gets a couple of files ... it does something with them, which involves computation (compression in this case), and it spits out a single file with the results (The compresses tar ball in this case). And since this interface is known already, users do not have a hard time to grasp how to use our scripts.

**Comment by dneise** [comment]
So we have two questions now, one technical (what package to use for command line parsing) and one design question. How do we want our colleagues to use our scripts?

**Comment by dneise** [comment]
Next up is the question about the number of output files. It is very easy for people to understand that `tar` (to again use that example) spits out one file ... the tar-ball. 

If a script spits out two files, then ... well how would you name that script? Can you explain people in a single sentence what that script does? I guess not. Often two files are a hint for doing two different things. 

But ... 

What if the *one* analysis contains a computationally expensive step, and *another* analysis contains the exact same step? Then combining the two analysis saves a lot of time.

So we should think about a "soft rule". Should we try to generally avoid multiple outputs? So that users can call scripts to get one specific output and use that for their purpose?

This includes plots. An analysis should not emit the analysis result **and** the plots.

I understand that: 
 1. during development of a script, one needs plots to understand if the analysis works correctly (hint: this does not replace unit tests for this analysis, and plots are difficult to use for unit tests, so better think of a more numeric test ... and don't forget to implement it ... it is easy to forget, once the plots looked good) and
 2. People need plots.

So .. this means .. an analysis should spit out the result. And a another tool should take this result and turn it into a plot. This increases speed, when it comes to publishing plots. As long as plotting is hardly wired with analysis, one always has to repeat the analysis when one wants to change the color of a line or histogram, or an axis title. That is a pity. 
I have often done this in my life ... made a plot ... showed it in a meeting .. got a question to modify the plot a bit ... answered: yes ... but that takes 10 hours. No! We should strive to become better than that.

So .. analysis scripts should try to write "enough" results into their output, so that people can generate plots from this ... possibly different plots, with different binning ... so if disk-space is not an issue, do not write binned data to disk.

So now we have kind of 3 questions:

 - should we try to strip "plotting" from "analysis" in the future? 
 - should we try to emit only one output file?
 - what command line parser should we use?

Maybe we can find more questions even, which are entangled with this matter.

**Comment by calispac** [comment]
> There are some scripts in digicampipe/ which cannot be used without modification, as for example they contain fixed paths into somebodies home folder. This should be avoided.

I totally agree. I propose we try to use the default paths from the test files. At least it should work out of the box and the user can have an example of what it should do. What do you think?

> what package to use for command line parsing

So far we have been using optparse (much like argpars I guess). If we have something that take advantage of the shell that could be great. 
How does ctapipe do it? What's your opinion on their command line interface?

**Comment by dneise** [comment]
optparse is deprecated, so we should not use that in 2017

https://docs.python.org/3.6/library/optparse.html

----

I personally like `docopt` a lot ... but I think it is not good for large projects, since it does not provide argument validation. 

ctapipe uses traitlets for their cli interface. But I would not recommend that at all.

The new hip thing seems to be `click`. 

Whatever we choose, it should be one of these: https://awesome-python.com/#command-line-tools
(simply because I think the people maintaining this page, only put stuff here, which is used by many people. And we want to use stuff which is has a large user base, so documentation is great, and many bugs have already been found by other people, so we do not have to run into them ... like in traitlets)

**Comment by calispac** [comment]
Also I suggest that we find something that is nicely integratred with a config file such as : .yml, json.
In principle it is just a parser to write I suppose but this might be already done by some CL libraries.

**Comment by dneise** [comment]
I would recommend to reflect over features ... they come at a cost.

E.g. config files can make usage more difficult, since options can now come from 3 places (in order of precendence):

 - defaults in the code itself (is overwritten by)
 - the config file (is overwritten by)
 - the command line

Now, if a user is not aware that there is a config file somewhere ... they might simply end up not knowing what the settings are. 

To stay in my former picture: `tar` does not need a config file. 

In order to decide if a certain feature is needed, I think what one needs are "user stories". By looking at user stories we can understand if config files are needed to support the features we need to make a certain user story possible, or if there might be a different (easier) way to achieve the same outcome. 

In case we go for config files, the package I've seen (but never used) are:

 - https://pypi.python.org/pypi/ConfigArgParse
 - and docopt has this little example, combine config files with docopt: 
  https://github.com/docopt/docopt/blob/master/examples/config_file_example.py

**Comment by calispac** [comment]
Just wanted to write down what @dneise pointed out to me recently. If one goes for `doctop` then there is `Schema` for argument validation : 

https://github.com/docopt/docopt/blob/master/examples/validation_example.py
https://pypi.python.org/pypi/schema

**Comment by dneise** [comment]
In my PR for digicamtoy.produce_data i used 

https://pypi.python.org/pypi/commandr/1.5.0

Seems to worl quite nicely for functions with a lot of parameters.
